### PR TITLE
[systeminfo] Migrate to BND

### PR DESCRIPTION
**Pull request by falkena**
Signed-off-by: Alexander Falkenstern <alexander.falkenstern@gmail.com>

### CODE REVIEW

File: bundles/org.openhab.binding.systeminfo/.classpath [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="con" path="org.eclipse.m2e.MAVEN2_CLASSPATH_CONTAINER">
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="lib" path="lib/Moka7-1.0.2_io_patch.jar"/>
```
Comment (davidgraeff) [inline bundles/org.openhab.binding.systeminfo/.classpath]: copy/paste error I guess. can be removed

File: bundles/org.openhab.binding.systeminfo/pom.xml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  <artifactId>org.openhab.binding.systeminfo</artifactId>
+
+  <name>openHAB Add-ons :: Bundles :: Systeminfo Binding</name>
+
+  <dependencies>
+    <dependency>
+      <groupId>com.github.oshi</groupId>
+      <artifactId>oshi-core</artifactId>
+      <version>3.13.0</version>
```
Comment (davidgraeff) [inline bundles/org.openhab.binding.systeminfo/pom.xml]: add `<scope>provided</scope>` to all of them

File: bundles/org.openhab.binding.systeminfo/.classpath (hunk 1 above)
Comment (falkena) [reply bundles/org.openhab.binding.systeminfo/.classpath]: Ohh man. Thanks :-)

File: bundles/org.openhab.binding.systeminfo/pom.xml (hunk 2 above)
Comment (falkena) [reply bundles/org.openhab.binding.systeminfo/pom.xml]: Done

File: features/karaf/openhab-addons/src/main/feature/feature.xml [hunk 3]
```diff
@@ -847,7 +847,10 @@
 
     <feature name="openhab-binding-systeminfo" description="System Info Binding" version="${project.version}">
         <feature>openhab-runtime-base</feature>
-        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.systeminfo/${project.version}</bundle>
+        <bundle dependency="true">wrap:mvn:com.github.oshi/oshi-core/3.13.0</bundle>
+        <bundle dependency="true">wrap:mvn:net.java.dev.jna/jna/5.2.0</bundle>
```
Comment (J-N-K) [inline features/karaf/openhab-addons/src/main/feature/feature.xml]: wrapping should not be necessary, this is already an OSGi bundle

File: features/karaf/openhab-addons/src/main/feature/feature.xml [hunk 4]
```diff
@@ -847,7 +847,10 @@
...
+        <bundle dependency="true">wrap:mvn:net.java.dev.jna/jna-platform/5.2.0</bundle>
```
Comment (J-N-K) [inline features/karaf/openhab-addons/src/main/feature/feature.xml]: wrapping should not be necessary, this is already an OSGi bundle

File: features/karaf/openhab-addons/src/main/feature/feature.xml [hunk 5]
```diff
@@ -847,7 +847,10 @@
 
     <feature name="openhab-binding-systeminfo" description="System Info Binding" version="${project.version}">
         <feature>openhab-runtime-base</feature>
-        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.systeminfo/${project.version}</bundle>
+        <bundle dependency="true">wrap:mvn:com.github.oshi/oshi-core/3.13.0</bundle>
```
Comment (J-N-K) [inline features/karaf/openhab-addons/src/main/feature/feature.xml]: wrapping should not be necessary, this is already an OSGi bundle

File: features/karaf/openhab-addons/src/main/feature/feature.xml (hunk 3 above)
Comment (falkena) [reply features/karaf/openhab-addons/src/main/feature/feature.xml]: Played around to solve travis issues. Sadly, still no progress.

File: features/karaf/openhab-addons/src/main/feature/feature.xml (hunk 4 above)
Comment (falkena) [reply features/karaf/openhab-addons/src/main/feature/feature.xml]: Played around to solve travis issues. Sadly, still no progress.

File: features/karaf/openhab-addons/src/main/feature/feature.xml (hunk 5 above)
Comment (falkena) [reply features/karaf/openhab-addons/src/main/feature/feature.xml]: Played around to solve travis issues. Sadly, still no progress.

File: features/karaf/openhab-addons/src/main/feature/feature.xml (hunk 4 above)
Comment (J-N-K) [reply features/karaf/openhab-addons/src/main/feature/feature.xml]: I don‘t think Travis failure is related to your changes. Travis fails on all PR.

**Comment by falkena** [comment]
I stuck with karaf feature verification: openHAB Add-ons :: Features :: Karaf :: Add-ons .... FAILURE
The message is 

> Feature resolution failed for [openhab-binding-systeminfo/2.5.0.SNAPSHOT]
> [ERROR] Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=openhab-binding-systeminfo; type=karaf.feature; version=2.5.0.SNAPSHOT; filter:="(&(osgi.identity=openhab-binding-systeminfo)(type=karaf.feature)(version>=2.5.0.SNAPSHOT))" [caused by: Unable to resolve openhab-binding-systeminfo/2.5.0.SNAPSHOT: missing requirement [openhab-binding-systeminfo/2.5.0.SNAPSHOT] osgi.identity; osgi.identity=org.openhab.binding.systeminfo; type=osgi.bundle; version="[2.5.0.201904152011,2.5.0.201904152011]"; resolution:=mandatory [caused by: Unable to resolve org.openhab.binding.systeminfo/2.5.0.201904152011: missing requirement [org.openhab.binding.systeminfo/2.5.0.201904152011] osgi.wiring.package; filter:="(&(osgi.wiring.package=oshi.hardware)(version>=3.13.0)(!(version>=4.0.0)))" [caused by: Unable to resolve com.github.oshi.oshi-core/3.13.0: missing requirement [com.github.oshi.oshi-core/3.13.0] osgi.wiring.package; filter:="(osgi.wiring.package=com.sun.jna.platform.linux)"]]]

Build runs smoothly. Any hint?

File: features/karaf/openhab-addons/src/main/feature/feature.xml (hunk 4 above)
Comment (falkena) [reply features/karaf/openhab-addons/src/main/feature/feature.xml]: My local build fails too :-)

**Comment by J-N-K** [comment]
Seems to be a bug in JNA 5.2.0: https://github.com/java-native-access/jna/issues/1054

Did this ever work under Linux?

**Comment by J-N-K** [comment]
You can try removing the dependency to `net.java.dev.jna:.jna-platform`, putting ` jna-platform-5.2.0.jar` to `/lib` and add the following lines to `pom.xml`.

```
<properties>
  <bnd.exportpackage>com.sun.jna.platform.*;version="5.2.0"</bnd.exportpackage>
</properties>
```

I did not try but that should export all packages.

**Comment by falkena** [comment]
Hi @J-N-K , thank you :-) I've traced the dependecies down to jna 4.2.1: still the same problem. It seems, that it's broken for the long time and nobody catched the bug. On Raspberry jna 5.2.0 runs as expected. Therefore it seems, that we must live with workaround.

**Comment by J-N-K** [comment]
The export works?

**Comment by falkena** [comment]
Build and verification was OK. Currently i rebase my other PR and test all the stuff.

**Comment by falkena** [comment]
Ok. Rebased, updated eclipse, wait till workspace build... And new fail... 

> openhab> bundle:restart 254
> Error executing command: Error restarting bundles:
>         Unable to start bundle 254: Could not resolve module: org.openhab.binding.systeminfo [254]
>   Unresolved requirement: Import-Package: com.sun.jna; version="[5.2.0,6.0.0)"

Ok, embed jna-5.2.0.jar the same way as jna-platform-5.2.0.jar. Rebuilt, new attempt...
> openhab> bundle:restart 254
> Error executing command: Error restarting bundles:
>         Unable to start bundle 254: Could not resolve module: org.openhab.binding.systeminfo [254]
>   Unresolved requirement: Import-Package: oshi; version="[3.13.0,4.0.0)"

Ok, embed oshi-core-3.13.0.jar the same way as jna-5.2.0.jar. Rebuilt, new attempt...
> openhab> bundle:restart 254
> Error executing command: Error restarting bundles:
>         Unable to start bundle 254: Could not resolve module: org.openhab.binding.systeminfo [254]
>   Unresolved requirement: Import-Package: com.sun.management

Built on Windows and run on raspberry...

**Comment by davidgraeff** [comment]
Did it work with the old buildsystem? If not we are not loosing anything if we merge this PR for now and keep investigating later

**Comment by J-N-K** [comment]
This is probably due to the missing dependencies which are defined in the feature.

File: bundles/org.openhab.binding.systeminfo/.classpath [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="src" output="target/test-classes" path="src/test/java">
+		<attributes>
+			<attribute name="optional" value="true"/>
+			<attribute name="maven.pomderived" value="true"/>
+			<attribute name="test" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="output" path="target/classes"/>
```
Comment (J-N-K) [inline bundles/org.openhab.binding.systeminfo/.classpath]: add jna-platform as lib (see e.g. allplay)

File: bundles/org.openhab.binding.systeminfo/pom.xml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>com.github.oshi</groupId>
+      <artifactId>oshi-core</artifactId>
+      <version>3.13.0</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
```
Comment (J-N-K) [inline bundles/org.openhab.binding.systeminfo/pom.xml]: remove this dependency (always provided if embedded)

File: features/karaf/openhab-addons/src/main/feature/feature.xml [hunk 8]
```diff
@@ -855,7 +855,10 @@
 
     <feature name="openhab-binding-systeminfo" description="System Info Binding" version="${project.version}">
         <feature>openhab-runtime-base</feature>
-        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.systeminfo/${project.version}</bundle>
+        <bundle dependency="true">mvn:com.github.oshi/oshi-core/3.13.0</bundle>
+        <bundle dependency="true">mvn:net.java.dev.jna/jna-platform/5.2.0</bundle>
```
Comment (J-N-K) [inline features/karaf/openhab-addons/src/main/feature/feature.xml]: remove feature dependency (always provided if embedded)

**Comment by falkena** [comment]
@davidgraeff : Yes it run bevore @J-N-K I already performed requested changes yesterday, just not pushed them, since it will not run...

**Comment by J-N-K** [comment]
As I said before: the feature resolution in your production setup will not work because it is using the old feature.xml. 

It should work if you install the missing bundles manually. Did you try that?

File: bundles/org.openhab.binding.systeminfo/.classpath (hunk 6 above)
Comment (falkena) [reply bundles/org.openhab.binding.systeminfo/.classpath]: Removed in current version

File: bundles/org.openhab.binding.systeminfo/pom.xml (hunk 7 above)
Comment (falkena) [reply bundles/org.openhab.binding.systeminfo/pom.xml]: Done

**Comment by falkena** [comment]
I run two environments: production (2.5.0M1) and development mit latest snapshot. The strange thing is, i get 
> Unresolved requirement: Import-Package: oshi; version="[3.13.0,4.0.0)"
if i try to load the binding, but according to oshi-core manifest file, it's provided by oshi-core.jar...

**Comment by J-N-K** [comment]
I raised a PR against your repo. With those fixes everything compiles and I verified that the binding is working on two systems (both Linux). The tests fail, though. Maybe you can have a look at that.

**Comment by J-N-K** [comment]
@falkena, thanks to @matthiasblaesing JNA 5.3.0  was released very fast. I have updated my PR to your repo, please merge that, the tests now run fine, too.

**Comment by falkena** [comment]
> 
> 
> @falkena, thanks to @matthiasblaesing JNA 5.3.0 was released very fast. I have updated my PR to your repo, please merge that, the tests now run fine, too.

Hi @J-N-K : I've merged your PR. Thank you for help! I'll rebase my another PR and test the stuff this week. If something will go wrong, i'll fix it. May be @dbwiddis can find some time and update the oshi dependency to JNA 5.3.0. Officially OSHI depends now still on JNA 5.2.0

**Comment by dbwiddis** [comment]
Hah. I’ve been waiting on JNA 5.3 and checking almost daily. But missed a day.

**Comment by J-N-K** [comment]
But it works with 5.3.0. so no need to hurry.

**Pull request merged**